### DISCUSSION TITLE
ICommand and CommandBase - why duplicated in modules

### DISCUSSION THREAD

**Discussion by radian7** [kgrzybek/modular-monolith-with-ddd#137] [Ideas] [answered]
Hi,

First of all, good job, very inspired.

I have following question : 
why [interface ICommand] and [abstract class CommandBase] is duplicated in modules ?
for example:
CompanyName.MyMeetings.Modules.UserAccess.Application.Contracts
and 
CompanyName.MyMeetings.Modules.Payments.Application.Contracts
?

Isn't better to move it to BuildingBlocks project ?

Thanks.

**Comment by AndreiGanichev** [comment, marked as answer]
Hi, @radian7.
I suggests the reason is following. Two of the main attributes of modularity [1] are **independence** and **well-defined interface**. ICommand, IQuery are part of a module's interface and they shouldn't be shared to keep and emphasize module's independence.

Robert Martin [2] uses a term "accidental duplication" and I guess it's exactly what we have here: contracts looks like duplicates accidentally - they are independent actually.

1. Kamil Grzybek. [Modular Monolith: A Primer](http://www.kamilgrzybek.com/design/modular-monolith-primer/)
1. Robert Martin. Clean Architecture. Chapter 7. SRP.

**Reply by kgrzybek** [reply]
Hi @radian7 

@AndreiGanichev explanation is correct! 

I can only add that if you put these interfaces in Building Blocks, you will allow the API to call a Command / Query that the module does not support. See my answer for the same question here: https://github.com/kgrzybek/modular-monolith-with-ddd/issues/12#issuecomment-534493239

Code duplication does not always mean DRY violation. I very like this article on that subject https://verraes.net/2014/08/dry-is-about-knowledge/

**Reply by AndreiGanichev** [reply]
>I very like this article on that subject

Thanks for the link! Could you advise some for other design principles? Principles like SOLID, KISS, YAGNI are easy to read but it's quite tricky to understand them properly. So it's good to have such articles to read by myself and to share knowledge with teammates.
